### ISSUE TITLE
Asterisk binding critical bug

### ISSUE THREAD

**Issue by davidhrbaty** [openhab/openhab#2021]
Hello, yesterday I want to add asterisk binding to my openhab server, but I have had problem with null pointer exception.

I can see, that openhab is successfully connected to asterisk:

```
Connected to Asterisk 1.6.2.9-2+squeeze10 currently running on gw (pid = 750)
Verbosity is at least 10
gw*CLI> manager show connected 
  Username         IP Address       Start       Elapsed     FileDes   HttpCnt   Read   Write
  openhab          10.0.20.18       1421922958  763         11        0         08191  08191
1 users connected.
gw*CLI> 
```

But after starting openhab server I have seen this problem:

```
11:35:57.429 [DEBUG] [i.internal.GenericItemProvider:334  ] - Start processing binding configuration of Item 'Incoming_Call (Type=SwitchItem, State=Uninitialized)' with 'AsteriskGenericBindingProvider' reader.
11:35:57.452 [WARN ] [AsteriskGenericBindingProvider:78   ] - bindingConfig is NULL (item=Incoming_Call (Type=SwitchItem, State=Uninitialized)) -> processing bindingConfig aborted!
11:35:57.466 [DEBUG] [i.internal.GenericItemProvider:334  ] - Start processing binding configuration of Item 'Active_Call (Type=CallItem, State=Uninitialized)' with 'AsteriskGenericBindingProvider' reader.
11:35:57.477 [WARN ] [AsteriskGenericBindingProvider:78   ] - bindingConfig is NULL (item=Active_Call (Type=CallItem, State=Uninitialized)) -> processing bindingConfig aborted!
11:35:57.488 [DEBUG] [i.internal.GenericItemProvider:334  ] - Start processing binding configuration of Item 'Active_Call (Type=CallItem, State=Uninitialized)' with 'AsteriskGenericBindingProvider' reader.
```

I also send my items configuration file:

```
Switch Incoming_Call    "Ringing"           (Phone)    { asterisk=active }
Call Active_Call    "Connected [to %1$s from %2$s]" (Phone)    { asterisk=active }
Call Active_Call    "Connected to [%s]"     (Phone)    { asterisk=active }
```

After make any change in items file I have had a lot of exceptions in openhab.log:

```
11:47:58.739 [INFO ] [c.internal.ModelRepositoryImpl:97   ] - Refreshing model 'default.items'
11:47:58.852 [DEBUG] [i.internal.GenericItemProvider:154  ] - Processing binding configs for items from model 'default.items'
11:47:58.876 [ERROR] [.o.m.c.i.folder.FolderObserver:105  ] - An unexpected exception has occured
java.lang.NullPointerException: null
    at org.openhab.model.item.binding.AbstractGenericBindingProvider.removeConfigurations(AbstractGenericBindingProvider.java:93) ~[na:na]
    at org.openhab.model.item.internal.GenericItemProvider.processBindingConfigsFromModel(GenericItemProvider.java:164) ~[na:na]
    at org.openhab.model.item.internal.GenericItemProvider.modelChanged(GenericItemProvider.java:379) ~[na:na]
    at org.openhab.model.core.internal.ModelRepositoryImpl.notifyListeners(ModelRepositoryImpl.java:158) ~[na:na]
    at org.openhab.model.core.internal.ModelRepositoryImpl.addOrRefreshModel(ModelRepositoryImpl.java:99) ~[na:na]
    at org.openhab.model.core.internal.folder.FolderObserver.checkFolder(FolderObserver.java:142) ~[na:na]
    at org.openhab.model.core.internal.folder.FolderObserver.run(FolderObserver.java:99) ~[na:na]
```

Solution for these exceptions are restart openhab server only without binding asterisk.
Thank you for your response.
##

**Comment by teichsta** [comment]
Hi,

did you observe any problem using the Binding? At first sight i would assume the NPE's are "just" Start-Order issues. The items were already read while the Binding has not been started properly. Can you confirm that the Binding functionality is broken?

Best, Thomas E.-E.

**Comment by davidhrbaty** [comment]
Hello Thomas,

Thank you for your answer. Yes, you were right, "just" start-order issues, but unfornately Binding is still unfunctional. I have cleaned all items and sitemap files. I left only relevant items to this binding in items files:

```
Switch Incoming_Call    "Ringing"           (Phone)    { asterisk=active }
Call Active_Call    "Connected [to %1$s from %2$s]" (Phone)    { asterisk=active }
Call Active_Call    "Connected to [%s]"     (Phone)    { asterisk=active }
```

and also in sitemap file:

```
sitemap default label="OpenHouse"
{
        Frame label="Phone" {
                Switch item=Incoming_Call label="In call"
                Text item=Active_Call
        }
}
```
I have had no error in openhab.log in debug mode. With tcpdump between asterisk and openhab server I still saw communication, bud no state change happened if I had a call in Asterisk. 
Do you have any advice where can be a problem?

**Comment by davidhrbaty** [comment]
I can confirm, that binding functionality is broken. I tried several types of configuration without result.

**Comment by Luvien** [comment]
Hello,

I get the same problem "BindingConfig is NULL (item=Incoming_Call (Type=SwitchItem, State=Uninitialized)) -> processing bindingConfig aborted!"

I'm using IncrediblePBX Asterisk 11, openhab 1.7 and AMI 1.3

**Comment by Luvien** [comment]
Hello,

I just tried with IncrediblePBX Asterisk 12 on Ubuntu, AMI 2.5.0, same error. Something is not working with the binding.

Anyone can help?

**Comment by tradiuz** [comment]
The asterisk binding only works with Asterisk 1.6 or 1.8, IIRC.

**Comment by Luvien** [comment]
Thanks!  Hummm my guess is I'll try it.  I'm trying to see it openhab os
working with Freepbx as a replacement for Asterisk.

I will try Asterisknow 5.2 as you can install asterisk 1.8

Many thanks for your answer!

**Comment by Luvien** [comment]
Hello again

I tries Asterisk NOW 1.8 with openhab 1.6 and 1.7 and I still get the null
binding.

Thanks for trying,

LJ

**Comment by Stormwind99** [comment]
I get same error with OpenHAB 1.8.3 and Asterisk 13.7.2 .

**Comment by Stormwind99** [comment]
I wonder if this pull request would fix the issue: https://github.com/openhab/openhab/pull/3059

**Comment by 9037568** [comment]
I built a new [jar file](https://github.com/openhab/openhab/files/596272/org.openhab.binding.asterisk-1.9.0-SNAPSHOT.zip) for the binding based on the changes in PR #3059 .

Hopefully one of the interested parties ( @davidhrbaty @Luvien @Stormwind99  ) can try this out and see if it fixes this issue?

**Comment by udo1toni** [comment]
As I do not have a build system, could you please offer a link to the compiled jar-file? I would like to test the binding (here Asterisk 13.6.0)

**Comment by 9037568** [comment]
There's already a link to the jar file there: https://github.com/openhab/openhab/files/596272/org.openhab.binding.asterisk-1.9.0-SNAPSHOT.zip

Is it not working?

**Comment by udo1toni** [comment]
Hmm... I must do something wrong... the .zip does contain a .jar file (named asterisk-java.jar)
I put this file to $openhab-directory/addons/ and renamed it to org.openhab.binding.asterisk-1.9.0-SNAPSHOT.jar. I put this to openhab.cfg:
```
asterisk:host=<fqdn of asterisk>
asterisk:username=openhab 
asterisk:password=<the password>
```
.items:
```
Call AstCall "verbunden [nach %1$s von %2$s]" <phone> {asterisk="active"}
```

of course in asterisk, there is a user named openhab:
```
asterisk*CLI> manager show user openhab

          username: openhab
            secret: <Set>
               ACL: yes
         read perm: call
        write perm: <none>
   displayconnects: yes
allowmultiplelogin: yes
         Variables:

asterisk*CLI> manager show settings

Global Settings:
----------------
  Manager (AMI):             Yes
  Web Manager (AMI/HTTP):    Yes
  TCP Bindaddress:           0.0.0.0:5039
  HTTP Timeout (minutes):    60
  TLS Enable:                No
  TLS Bindaddress:           Disabled
  TLS Certfile:              asterisk.pem
  TLS Privatekey:
  TLS Cipher:
  Allow multiple login:      Yes
  Display connects:          Yes
  Timestamp events:          No
  Channel vars:
  Debug:                     No

asterisk*CLI> manager show connected
  Username         IP Address                                               Start       Elapsed     FileDes   HttpCnt   Read   Write
0 users connected.
```

Do I have to do something in addition?

**Comment by 9037568** [comment]
No, just rename the file from "org.openhab.binding.asterisk-1.9.0-SNAPSHOT.zip" to "org.openhab.binding.asterisk-1.9.0-SNAPSHOT.jar". No need to open it up...

**Comment by udo1toni** [comment]
Ah :) 

Now getting an error line:
```
2016-11-28 01:09:49.120 [ERROR] [o.b.a.internal.AsteriskBinding] - Could not connect to ManagerInterface on <fqdn>: java.net.ConnectException: Verbindungsaufbau abgelehnt
```
Do I have to setup the port?

**Comment by udo1toni** [comment]
Yes... port was wrong (accidently set up to 5039, but has to be 5038 in asterisk) now a step forward :)
```
2016-11-28 01:25:06.113 [WARN ] [internal.ManagerConnectionImpl] - Unsupported protocol version 'Asterisk Call Manager/2.8.0'. Use at your own risk.
```
But will test tomorrow, erm... later, today... ;)

**Comment by udo1toni** [comment]
Works as expected :) got the caller id and the called id. Here asterisk 13.6.0 (AMI 2.8.0)

**Comment by 9037568** [comment]
Fixed via merged PR #3059.
